foundation-models/openai/src/main/java/com/sap/ai/sdk/foundationmodels/openai/OpenAiStreamingHandler.java

```diff
@@ -0,0 +1,98 @@
+
+  @Nonnull private final Class<D> deltaType;
+
+  /**
+   * Processes a {@link ClassicHttpResponse} and returns some value corresponding to that response.
+   *
+   * @param response The response to process
+   * @return A {@link Stream} of a model class instantiated from the response
+   * @throws OpenAiClientException in case of a problem or the connection was aborted
+   */
+  @Nonnull
+  public Stream<D> handleResponse(@Nonnull final ClassicHttpResponse response)
+      throws OpenAiClientException {
+    if (response.getCode() >= 300) {
+      buildExceptionAndThrow(response);
+    }
+    return parseResponse(response);
+  }
+
+  /**
+   * @param response The response to process
+   * @return A {@link Stream} of a model class instantiated from the response
+   * @author stippi
+   */
+  // The stream is closed by the user of the Stream
+  @SuppressWarnings("PMD.CloseResource")
+  private Stream<D> parseResponse(@Nonnull final ClassicHttpResponse response)
+      throws OpenAiClientException {
+    final HttpEntity responseEntity = response.getEntity();
+    if (responseEntity == null) {
+      throw new OpenAiClientException("Response from OpenAI model was empty.");
+    }
+    final InputStream inputStream;
+    try {
+      inputStream = responseEntity.getContent();
+    } catch (IOException e) {
+      throw new OpenAiClientException("Failed to read response content.", e);
+    }
+    final var br = new BufferedReader(new InputStreamReader(inputStream, StandardCharsets.UTF_8));
+
+    // https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events/Using_server-sent_events#event_stream_format
+    return br.lines()
+        .filter(
+            responseLine ->
+                // half of the lines are empty newlines, the last line is "data: [DONE]"
+                !responseLine.isEmpty() && !"data: [DONE]".equals(responseLine.trim()))
+        .peek(
+            responseLine -> {
+              if (!responseLine.startsWith("data: ")) {
+                parseErrorAndThrow(
+                    responseLine,
+                    new OpenAiClientException("Failed to parse response from OpenAI model"));
+              }
+            })
+        .map(
+            responseLine -> {
+              final String data = responseLine.substring(5).replace("delta", "message");
+              try {
+                return JACKSON.readValue(data, deltaType);
+              } catch (final IOException e) {
+                log.error(
+                    "Failed to parse the following response from OpenAI model: {}", responseLine);
+                throw new OpenAiClientException(
+                    "Failed to parse delta message: " + responseLine, e);
+              }
+            })
+        .onClose(
+            () -> {
+              try {
+                inputStream.close();
+              } catch (IOException e) {
+                log.error("Could not close HTTP input stream", e);
+              }
+            });
```

Comment:
I'd be curious if this code is actually needed, or if the stream is also closed without it.

Reply:
This is actually needed, I tested without it and the test fails
_Note: the stream is only closed by using try-with-resources. consuming the stream fully (terminal operation) does not close it._

Reply:
> consuming the stream fully (terminal operation) does not close it.

Hmm, but if that is the case, how does adding an implementation for `.onClose` change anything?

Reply:
Because try-with-resources closes the stream

Reply:
ah okay, and the input stream is not closed without this? I would have thought that maybe whatever implementation of stream is the default used under the hood here would "pass on" the close to any underlying streams the final stream is built upon

But if we have to manually close it, why not close the buffered reader instead? [It will close the input stream internally](https://stackoverflow.com/questions/1388602/do-i-need-to-close-both-filereader-and-bufferedreader).